File: advent-of-code-2024/src/day_01.rs
use std::fs;
use std::time::Instant;

// https://adventofcode.com/2024/day/1

const DAY_STRING : &str = "day_01";
const TEST_RUNS : u32 = 1000;

fn part_1() -> String
{
    let input = get_input(false);

    let mut left_numbers : Vec<i32>;
    let mut right_numbers : Vec<i32>;

    (left_numbers, right_numbers) = split_input_into_two_collections(input);

    left_numbers.sort();
    right_numbers.sort();

    let mut distance : i32 = 0;
    for i in 0..left_numbers.len() {
        distance += (right_numbers[i] - left_numbers[i]).abs();
    }

    return distance.to_string();
}

fn part_2() -> String
{
    let input = get_input(false);

    let left_numbers : Vec<i32>;
    let right_numbers : Vec<i32>;

    (left_numbers, right_numbers) = split_input_into_two_collections(input);

    let mut similarity : i32 = 0;
    for i in 0..left_numbers.len() {
        let count : i32 = right_numbers.iter().filter(|&n| *n == left_numbers[i]).count().try_into().unwrap();
        similarity += left_numbers[i] * count;
    }

    return similarity.to_string();
}

pub fn solve() {
    let mut before = Instant::now();

    let part_1_result = part_1();
    for _ in 1..TEST_RUNS {
        part_1();
    }
    let part_1_time_elapsed = before.elapsed() / TEST_RUNS;

    before = Instant::now();

    let part_2_result = part_2();
    for _ in 1..TEST_RUNS {
        part_2();
    }
    let part_2_time_elapsed = before.elapsed() / TEST_RUNS;

    println!("\n-- Day 01 --");
    println!(" Part 1: {:15} elapsed: {:.2?} (average of 1000 runs)", part_1_result, part_1_time_elapsed);
    println!(" Part 2: {:15} elapsed: {:.2?} (average of 1000 runs)", part_2_result, part_2_time_elapsed);
}

fn get_input(use_test_data : bool) -> Vec<String> {
    let file_path = match use_test_data {
        true => format!("test_input/{}.txt", DAY_STRING),
        false => format!("input/{}.txt", DAY_STRING)
    };

    let data : Vec<String> = fs::read_to_string(file_path)
        .unwrap()
        .lines()
        .map(String::from)
        .collect();

    return data;
}

fn split_input_into_two_collections(input : Vec<String>) -> (Vec<i32>, Vec<i32>) {
    let mut left_numbers : Vec<i32> = vec![];
    let mut right_numbers : Vec<i32> = vec![];

    for line in input {
        let parts : Vec<String> = line.trim().split_whitespace().map(String::from).collect();

        let left : i32 = parts[0].parse()
            .expect("Failed to parse");
        let right : i32 = parts[1].parse()
            .expect("Failed to parse");

        left_numbers.push(left);
        right_numbers.push(right);
    }

    return (left_numbers, right_numbers)
}

File: advent-of-code-2025/day_01.py
def part1(data):
	point_to_zero_counter = 0
	current_position = 50

	for line in data:
		direction = line[0]
		steps = int(line[1:])
		steps = steps % 100

		if direction == "L":
			current_position -= steps
			if current_position < 0:
				current_position += 100
		else:
			current_position += steps
			if current_position > 99:
				current_position -= 100

		if current_position == 0:
				point_to_zero_counter += 1

	print("Part 1: ", point_to_zero_counter)

def part2(data):
	point_to_zero_counter = 0
	current_position = 50
	at_zero = False

	for line in data:
		direction = line[0]
		steps = int(line[1:])

		point_to_zero_counter += steps // 100
		steps = steps % 100

		if direction == "L":
			current_position -= steps

			if not at_zero and current_position < 0:
				point_to_zero_counter += 1
			if current_position < 0:
				current_position += 100
		else:
			current_position += steps
			if current_position > 100:
				point_to_zero_counter += 1

			if current_position > 99:
				current_position -= 100

		if current_position == 0:
			point_to_zero_counter += 1
			at_zero = True
		else:
			at_zero = False

	print("Part 2: ", point_to_zero_counter)

with open("input/day_01.txt", 'r') as file:
	data = file.read().splitlines()
	part1(data)
	part2(data)